utility/src/main/java/org/oppia/android/util/networking/NetworkConnectionUtil.kt

```diff
@@ -11,7 +11,10 @@ interface NetworkConnectionUtil {
     CELLULAR,
 
     /** Not connected to a network. */
-    NONE
+    NONE,
+
+    /** Refers to the actual connection status of the device. */
+    DEFAULT
```

Comment:
Does this make sense in the production sense of the component? It seems like a debug-only property. Could we redesign this such that 'default' is only usable in debug situations?

Reply:
Not really. But putting this in `NetworkConnectionUtilDebugImpl` will require to import it when we are passing DEFAULT connection status to the `ForceNewtorkTypeViewModel` in #3502. An alternative could be to go back to the **null** approach instead of setting **forcedConnectionStatus** as DEFAULT. WDYT?

Reply:
Hmm. ``null`` has the same problem, I think, it just changes what the sentinel value is in this case. I actually think that importing the debug implementation _is_ the correct way to do this; the purpose of the debug version is to be used in both production and debug cases (in the same way that a test implementation is used in prod and tests).

That being said, depending directly on a debug impl isn't so great. It may actually be cleaner to have something like NetworkConnectionDebugUtil that exposes whatever API you need in debug-only cases and then have the implementation implement both the prod & debug interfaces (& have corresponding bindings so that the view model can inject the debug version when it needs to).

This, I think, would be a relatively small change. Please reach out if you have any concerns or if I didn't explain this well.

Reply:
This seems a good approach but I am facing some issues with this. Let me describe.

There are two interfaces now: `NetworkConnectionUtil` and `NetworkConnectionDebugUtil`.

`NetworkConnectionUtilProdImpl` implements `NetworkConnectionUtil` whose binding is provided by `NetworkConnectionUtilProdModule`. 

`NetworkConnectionUtilDebugImpl` implements both `NetworkConnectionUtil` and `NetworkConnectionDebugUtil` whose corresponding bindings are provided by `NetworkConnectionUtilDebugModule`. 

`ForceNetworkTypeFragmentPresenter` will be injecting `NetworkConnectionDebugUtil` as it requires debug-only functions i.e, _setNetworkConnectionStatus_ and _getForcedConnectionStatus_.

Now, when we are building the app in debug mode (i.e, putting `NetworkConnectionUtilDebugModule` in `ApplicationComponent`) then everything is fine. But when we are building the app in prod mode (i.e, putting `NetworkConnectionUtilProdModule` in `ApplicationComponent`), the Gradle build is failing as it cannot provide dependencies for `NetworkConnectionDebugUtil` which is used in `ForceNetworkTypeFragmentPresenter`.

Can you think of a way to resolve this?

A quick alternative could be to use `NetworkConnectionUtilDebugImpl` only instead of creating a debug util interface. Or, we can leave the DEFAULT case in `NetworkConnectionUtil` only as you mentioned debug version can be used in both production and debug cases (I'm not sure if this is what you meant exactly).

WDYT about this @BenHenning?

Reply:
Ah, another case of Gradle making this sort of dependency separation more difficult.

I think our only option is to introduce a BindsOptionalOf for the debug interface, and the inject an Optional<NetworkConnectionDebugUtil> (see 'Optional bindings' under https://dagger.dev/dev-guide/). If we go this route, please add a TODO to the optional binding to remove it once we move to Bazel (a TODO on #59 should be sufficient for now).

Would this work?

Reply:
I have used optional binding now and it works fine in both builds. I have also added the TODO. PTAL. Thanks.